### PR TITLE
Add initial version of Function::from_str()

### PR DESCRIPTION
**Pull request by eloylp**
This an initial version, it probably could be more resilient/optimized.

We can add more test cases easily in the future.

This is still a work in progress.

### CODE REVIEW

File: uniffi_docs/src/lib.rs [hunk 1]
```diff
@@ -2,20 +2,110 @@
...
 }
 
+impl FromStr for Function {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        // Converting comment into markdown first.
+        let md = s.replace("/// ", "").replace("///", "");
+
+        // Prepare the different buffers for the parse.
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:27]: Let's avoid inline comments that are somewhat obvious from the code alone.

File: uniffi_docs/src/lib.rs [hunk 2]
```diff
@@ -2,20 +2,110 @@
...
+        let mut description_buff = String::new();
+
+        let mut args_values_buff: Vec<String> = Vec::new();
+        let mut args_keys_buff: Vec<String> = Vec::new();
+
+        let mut return_description_buff = String::new();
+
+        // The current parsing stage.
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:35]: Let's avoid inline comments that are somewhat obvious from the code alone.

File: uniffi_docs/src/lib.rs [hunk 3]
```diff
@@ -2,20 +2,110 @@
...
+        let mut stage = ParseStage::DESCRIPTION;
+
+        // A new parser
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:38]: Let's avoid inline comments that are somewhat obvious from the code alone.

File: uniffi_docs/src/lib.rs [hunk 4]
```diff
@@ -144,3 +227,50 @@ pub fn extract_documentation(path: &Utf8Path) -> Result<Documentation> {
         impls,
     })
 }
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn it_parses_a_md_description() {
+        let description = std::fs::read_to_string("./tests/parsable_description.txt").unwrap();
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:236]: IMO, just inline that text in the test itself - it's not that long, and it's better to not have to jump around files to get the whole picture of the unit test (when it's possible).

You can use [indoc](https://docs.rs/indoc/latest/indoc/) for that (to be able to have pretty indent without affecting the input).

File: uniffi_docs/src/lib.rs [hunk 5]
```diff
@@ -144,3 +227,50 @@ pub fn extract_documentation(path: &Utf8Path) -> Result<Documentation> {
         impls,
     })
 }
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn it_parses_a_md_description() {
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:235]: In Rust people tend to name tests in form `test_some_test` (although it's not like a forced convention or something).

See: https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html
Or: https://github.com/rust-random/rand/blob/dbbc1bf3176138c867f3d84c0c4d288119a5a84e/rand_distr/src/gumbel.rs#L91-L160

File: uniffi_docs/src/lib.rs [hunk 6]
```diff
@@ -103,16 +193,9 @@ pub fn extract_documentation(path: &Utf8Path) -> Result<Documentation> {
                                     None
                                 }
                             })
+                            // TODO.Currently unwrap, as we do not expect errors. Just being compliant with API. Better handling.
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:196]: Ok. I agree, let's remove the comment.

File: uniffi_docs/src/lib.rs [hunk 7]
```diff
@@ -2,20 +2,110 @@
...
+        Ok(Function {
+            description: description_buff,
+            arguments_descriptions,
+            return_description,
+        })
+    }
+}
+
+enum ParseStage {
+    DESCRIPTION,
```
Comment (zduny) [inline uniffi_docs/src/lib.rs:105]: Enum variants should be named using CamelCase.
See: https://doc.rust-lang.org/rust-by-example/custom_types/enum.html

**Review by zduny — CHANGES_REQUESTED** [review]

**Comment by zduny** [comment]
It looks good overall, just resolve the comments.

File: uniffi_docs/src/lib.rs [hunk 8]
```diff
@@ -144,3 +227,50 @@ pub fn extract_documentation(path: &Utf8Path) -> Result<Documentation> {
...
+            return_description: Some(
+                "This is return value description.\nHere is a second line.\n".to_string(),
+            ),
+        }
+    }
+
+    #[test]
+    fn it_parses_a_no_md_description() {
+        let description = std::fs::read_to_string("./tests/no_md_description.txt").unwrap();
+        let cleaned_description = description.replace("/// ", "").replace("///", "");
```
[the remark this answers is not in the document]
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: No prob ! im glad to remove such code :smile: . Fixed at f3c35500882a15eac85ac24a9501dc7ba976b07c .

File: uniffi_docs/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Agreed ! Loving `indoc` crate BTW. Fixed at f2f7c913cf05f0033fe749e37b62d59edfa40a81 .

File: uniffi_docs/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Fixed at e33d469eb595801c5477968d848a7a0ede9d9de0 .

File: uniffi_docs/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Fixed at e33d469eb595801c5477968d848a7a0ede9d9de0 .

File: uniffi_docs/src/lib.rs [hunk 9]
```diff
@@ -2,20 +2,110 @@
...
+
+        Ok(Function {
+            description: description_buff,
+            arguments_descriptions,
+            return_description,
+        })
+    }
+}
+
+enum ParseStage {
```
[the remark this answers is not in the document]
Comment (eloylp) [reply uniffi_docs/src/lib.rs:99]: Fixed at 5c2bcec84f8224b7d986e4c44dde012f1d015565.

File: uniffi_docs/src/lib.rs (hunk 5 above)
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Not sure where i read the other nomenclature. I prefer the one you mentioned `test_some_test` . Fixed at 03418a9043dbcfb34498ff79da5502763e1b1bb0 .

File: uniffi_docs/src/lib.rs (hunk 6 above)
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Fixed at e33d469eb595801c5477968d848a7a0ede9d9de0.

File: uniffi_docs/src/lib.rs (hunk 7 above)
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Nice catch. Just passed clippy and warned me about the issue. Fixed at af2117d2b711d7defa34cbb9686baf47fcbc3d25 .

**Comment by eloylp** [comment]
> It looks good overall, just resolve the comments.

Thanks for the detailed review @zduny . Fixed all the problems.

If at some point we need more resiliency in the function itself i think we can improve it later (ondemand).

Moving to "ready for review".

File: uniffi_docs/src/lib.rs [hunk 10]
```diff
@@ -2,20 +2,110 @@
...
     pub description: String,
     pub arguments_descriptions: HashMap<String, String>,
     pub return_description: Option<String>,
 }
 
+impl FromStr for Function {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        // Converting comment into markdown first.
```
[the remark this answers is not in the document]
Comment (eloylp) [reply uniffi_docs/src/lib.rs]: Fixed at https://github.com/eigerco/uniffi-rs/commit/f3c35500882a15eac85ac24a9501dc7ba976b07c .

**Review by zduny — APPROVED** [review]

**Pull request merged**